### ISSUE TITLE
Service Worker: Hash Mismatch

### ISSUE THREAD

**Issue by manfredsteyer** [angular/angular#23613]
I'd like to use this to start a discussion based upon a topic I talked about with @gkalpak at ng-conf 2018.

As mentioned in #21288, the service worker implementation checkes the hashes of the cached files and if they don't match, it just serves the data from the network (which is like having no service worker).

The problem is that every proxy on the web can manipulate the files downloaded. This is especially the case for mobile providers that are minifying and inlining a lot of stuff on the fly to save bandwith. But also tools like live-server change the index.html (and so they can be used to reproduce this issue).

As one of the big use cases for PWAs are areas with low bandwith, this is somehow conflicting.

Perhaps one way to solve this is to provide an exchangeable HashCheckStrategy. In cases where I as the programmer wants to take the responsiblity for not checking the hashes in order to bypass this issue, I could write sth like this:

```
export class BruceWillisHashCheckStrategy implements HashCheckStrategy {
     check(currentHash, expectedHash, fileName) {
         return true;
     }
}
```

**Comment by gkalpak** [comment]
Yes, I've been bitten by this myself and I suspect many others have been/will be too (with or without realizing :grin:). It is hard to track down and it would be great if we could come up with an easy to use solution that remained reliable.

##
My first thought was to be able to automatically put the hash of a file as a comment (e.g. at the top of the file), so that the SW could read the hash from the file (instead of hashing it itself). This might not work for certain types of files though (e.g. images).

##
Having a configurable `HashCheckStrategy` would be nice _iff_ we could ensure that it is a reliable one (which I don't think we can). Being able to reliably determining file content equality (which is what hash comparison gives us) is essential to the SW's working as expected (and not serving incorrect content). 

Giving users the power to easily mess this fundamental check is very dangerous imo. (E.g. with the "pass all" strategy shown above, you could easily end up being served an old, outdated asset (image, stylesheet, script) along with the latest `index.html` and app files, with unexpected results.

##
@alxhub, any thoughts on that? Do you know/remember what other strategies (other than comparing hashes) have been considered/rejected?

**Comment by D2KX** [comment]
I have run into similar issues when combining SSR (Universal) and Service Workers, as the pre-rendered index.html hash differs from the one in the manifest, making offline support impossible. I currently use the work-around to remove the hash from the index.html before deploying.

**Comment by vytautas-pranskunas-** [comment]
@D2KX can you sare some code snippet how you remove it or explain it in more deatils?

**Comment by gkalpak** [comment]
Hm...actually, I just tried this is worked fine:
1. Download the [finished sample code][1] from the [Universal Guide][2].
2. Unzipped and installed dependencies (with `npm install`).
3. Enabled SW with `ng add @angular/pwa --project=angular.io-example`.
4. Built with `npm run build:ssr`.
5. Run with `npm run serve:ssr`.

SSR works as expected on the first request (before the SW is activated). Once the SW takes over, the non-prerendered `index.html` is served, but that is OK, since it is served by the SW and doesn't require a round-trip to the server. (Offline mode also works fine.)

So, SSR speeds up the first request and SW speeds up subsequent requests (and enables offline mode).

##
So, that usecase works fine. Using `app-shell` is a different story. I'll have to investigate that more :smiley:

[1]: https://next.angular.io/generated/zips/universal/universal.zip
[2]: https://next.angular.io/guide/universal

**Comment by gkalpak** [comment]
> Regarding https://github.com/angular/angular/issues/23613#issuecomment-403527623

SEO is not affected, since spiders don't use ServiceWorkers.

**Comment by vytautas-pranskunas-** [comment]
I have tried it and it works however favicon is not cached for some reason and fails on offline mode
![image](https://user-images.githubusercontent.com/3918999/42510518-777deada-844f-11e8-8f0d-918de75240cc.png)

**Comment by gkalpak** [comment]
This is unrelated to this issue. If you think it is a bug in Angular, please open a new issue (providing the necessary details, including a reproduction).

But it sounds more like a support request, for which [StackOverflow](http://stackoverflow.com/) (using tag `angular`) is a more suitable place.

**Comment by vytautas-pranskunas-** [comment]
But you said you have tried it did'nt you notice that? :)

**Comment by jvitor83** [comment]
I cound't more agree with @manfredsteyer
The hash strategy is not a suitable one for much reasons (he described some). I really would like that the strategy could be configured.
Already got with that problem at https://github.com/angular/angular/issues/25307

**Comment by jvitor83** [comment]
I believe that allow some strategy focused in version is a more suited one.
Ex: If your app is 1.0 _(`package.json` maybe?)_ in source code and production, then you update the version (`package.json`) and build the files, somehow _(maybe the version of package.json go to some json file)_ the sw should check for this version and if not equals then allow the update.

What you think @gkalpak @alxhub ?

**Comment by jvitor83** [comment]
Or even to use timestamp instead of version number.
So, each build has a new timestamp and update if the cached one is different.

**Comment by jvitor83** [comment]
As a workaround, i think we can use the `sed -i 's/fetchedHash !== canonicalHash/false/g' ngsw-worker.js` after build, which change the line https://github.com/angular/angular/blob/master/packages/service-worker/worker/src/assets.ts#L399
from 
```
makeCacheBustedRequest = (fetchedHash !== canonicalHash);
```
to
```
makeCacheBustedRequest = (false);
```

By some testing, it will still work offline and everything. The only down side is that it will trigger the `this.swUpdate.available` event every time the hash change, consequently if you notify the user about the new version, it will show this message at every new hash (caused by external events).
_Keep in mind that the user can been notified about some new version and the version/app is the same! 
But i think it's better than keeping the user with some old version (as @gkalpak said it happen https://github.com/angular/angular/issues/25307#issuecomment-410632977)!_

Even with that workaround, the dev can check at update event if is really some new version of the app or any other thing that change the hash by checking some version variable that exists inside the code.

Can @gkalpak or @alxhub see if there are any side-effect i am not seeing?

**Comment by jvitor83** [comment]
Ok @gkalpak, thanks for the info.
Maybe some json file with version or timestamp was not a good solution, but if we put that at the service worker itself?

The service-worker keep the files at cache with the timestamp as key (`new Date().getTime()`). _At build-time the line `const CACHE_VERSION = '1535211072719';` is added to it._

_Just to clarify the idea, i will put some sw and sw-build._

**sw-build.js** (`node sw-build.js`)
```
var fs = require('fs');
var t = new Date().getTime();

var file = 'dist/service-worker.js';
var data = fs.readFileSync(file); //read existing contents into data
var fd = fs.openSync(file, 'w+');

var buffer = new Buffer("const CACHE_VERSION = '"+t+"';");
fs.writeSync(fd, buffer, 0, buffer.length, 0); //write new data
fs.writeSync(fd, data, 0, data.length, buffer.length);
fs.close(fd);
```

**service-worker.js**
```
// tick this to make the cache invalidate and update
//CACHE_VERSION Will be added at build time.
//const CACHE_VERSION = '';
const CURRENT_CACHES = {
  'read-through': 'read-through-cache-v' + CACHE_VERSION
};

self.addEventListener('activate', (event) => {
  const expectedCacheNames = Object.keys(CURRENT_CACHES).map((key) => {
    return CURRENT_CACHES[key];
  });

  event.waitUntil(
    caches.keys().then((cacheNames) => {
      return Promise.all(
        cacheNames.map((cacheName) => {
          if (expectedCacheNames.indexOf(cacheName) === -1) {
            // If this cache name isn't present in the array of "expected" cache names, then delete it.
            console.log('Deleting out of date cache:', cacheName);
            return caches.delete(cacheName);
          }
        })
      );
    })
  );
});

self.addEventListener('fetch', (event) => {

  event.respondWith(
    caches.open(CURRENT_CACHES['read-through']).then((cache) => {
      return cache.match(event.request).then((response) => {
        if (response) {
          // If there is an entry in the cache for event.request, then response will be defined
          // and we can just return it.
          return response;
        }

        console.log(' No response for %s found in cache. ' +
          'About to fetch from network...', event.request.url);

        return fetch(event.request.clone()).then((response) => {

          // Optional: add in extra conditions here, e.g. response.type == 'basic' to only cache
          // responses from the same domain. See https://fetch.spec.whatwg.org/#concept-response-type
          if (response.status < 400 && response.type === 'basic') {
            cache.put(event.request, response.clone());
          }
          return response;
        });
      }).catch((error) => {
        console.error('  Read-through caching failed:', error);
        throw error;
      });
    })
  );
});
```

**Comment by gkalpak** [comment]
Not sure what problem your suggestion addresses. First of all, the SW implementation (`ngsw-worker.js`) is independent of the app content (served or cached files). Furthermore, the SW is already "stamping" each app version by the SHA1 hash of its manifest file (which is turn dependeds on the `ngsw-config` plus file contents).

**Comment by jvitor83** [comment]
Well, my question is: "why do we need to hash each file? why check if it changes (by hash comparison)?"

If it is to "not download unnecessary file/asset _(by having the same hash)_ to avoid use the network _(because of potentially bad connection)_", i guess in this scenarios (where hash is changed by external events) this 'feature' can't be considered, because the hash is not reliable. Maybe, whatever the reason why the hash is compared lost it's purpose in this context.

What i trying to see is "the sw can't just check if is a new app version, erase the cache and **get the new files without the hash comparison**?"
_This can't be something the dev should choose if he fits that scenario?_

Or maybe i am missing something important?

**Comment by gkalpak** [comment]
One reason is to conserve bandwidth (which might be significant if you cache large, rarely-chaning files, like images, etc.)

##
But that is not the main reasons (as far as I understand).

There is the concept of an app version. An app version contains all the files that comprise the app. This info is captured in a "manifest" file (`ngsw.json`), which is used by the SW to find out about a new app version.

Note that the SW may serve different app versions to different app instances (e.g. different tabs). The SW guarrantees that each app instance will continue to receive the files that correspond to its assigned version. This is important, because if your user has the app open in tab 1, and you deploy a new version with an updated asset (which might be incompatible with the old app version - e.g. a new JS script that expects the new `index.html` layout and break if used with the old one), you don't want the user to suddenly start getting the new incompatible asset while still using the old `index.html`.

Therefore, it is important for the SW to be able to tell if an asset file belongs to a specific version. The only reliable way discussed so far, is comparing the content (which btw is also what browsers do for detecting modified ServiceWorker scripts).

For example, imagine you deploy version 2 which includes `foo.js` as an asset. The SW detects the new version and starts downloading the assets for that version. In the meantime, you deploy version 3, with a modified `foo.js`. SW gets the new (backwards incompatible) `foo.js`. If it were to happily cache and serve that version with the rest of the app files for version 2, your app would be broken.

The SW needs to be able to tell whether a received file is actually the one that belongs to the current version it is updating to. Also, lazily fetched assets (with `installMode: 'lazy'`) complicate things even more.

##
I am sure there are other usecases, but in general the idea is that the SW needs to be able to tell whether the files that it receives, when fetching for a specific app version, do indeed correspond to the files that where on disk when that version's build was created.

Verifying the content (via a hash) is the only reliable way that has come up :smiley:

**Comment by webmaxru** [comment]
@gkalpak That description deserves to be a part of the official tech explainer. Great job, mate!

**Comment by jvitor83** [comment]
So, what we have by now if the hash mismatch happen is:

1 - (CURRENTLY) Keep the user stuck with some old version (without risk broke/corrupt the app) and try to resolve what is causing the mismatch (in some cases could be even impossible)

2 - Allow hash mismatch, but when happen, clear the cache and update all clients/tabs (to avoid broke/corrupt the app - _but with the down side of refresh at each new version or hash mismatch_)

3 - Put the hash as comment in each file (not work on all file types), but can at least avoid the mismatch of text files

Thanks @gkalpak for your time to clarify the behaviour/issue.

**Comment by jvitor83** [comment]
Ok guys, after some strugglin i found and fix my hash mismatch problem. (This is not the solution to this issue anyway)
I will put it here to help others who can have the same problem.

---------------

My infrastructure topology is:
User -> Firewall -> Reverse Proxy -> Reverse Proxy (Ingress - Kubernetes NGINX) -> Web Server (NGINX)

What was happening is that the first Reverse Proxy was using some configuration that rewrite automatically links from http to https.
In my case, i am using some css lib that has commented http links to its page like:
```css
/*
  My CSS lib: http://some.random.page/
*/
```
Then in this case, the reverse proxy was **changing the links on-the-fly** (so the content was changed) and results in a hash mismatch.

Once this configuration was disabled, the hash mismatch issue goes away and the app updates successfully.

--------------

I think that in some cases this issue will be a very hard thing to bypass (identify/solve) or even impossible (where it is hosted in a environment out of devs control or unknow).

Anyway, thanks again for @gkalpak for the time to clarify the behaviour/issue.

**Comment by gkalpak** [comment]
Wow, good job finding that sneaky culprit 💯
And thx for posting your findings for other people to benefit from your work 👍

I think it might be worth mentioning in the docs that for the SW to work the build artifacts have to be served unmodified to the end-user. If anyone feels like submitting a pull request that add something like that in the docs, that would be awesome!

**Comment by raza2022** [comment]
@gkalpak how can we stop SW as I want to always want to get  the prerendered `index.html` each time but i got it (prerendered) only first time

**Comment by gkalpak** [comment]
Generally, the idea of prerendering is that it kicks in on the first load (to speed up the first rendering) and on subsequent loads the SW takes over and speeds up the rendering by serving the HTML and other resources from the local cache (and only fetching the necessary data).

If you wanted to prevent the SW from caching `index.html` (assuming you had a valid reason - I can't think of one 😛), you could remove it from any asset-group in `ngsw-config.json`.

**Comment by DanielHabenicht** [comment]
For everybody looking for a straightforward way to update the hashes in the `ngsw.json` I wrote this little script: 
```bash
#!/bin/bash

# This updates the hash of the index.html in the corresponding ngsw.json 
# Execute it after replacing the environment variables

# Paths to folders containing the index.html files (where the hash should be recalculated)
paths=( "de" "en")

# The first parameter has to be the path to the directory with the language folders
if [[ -z $1 ]]; then
    echo 'ERROR: No target file or directory given.'
    exit 1
fi

# Go to all folders containing index.html and ngsw.json
for i in "${paths[@]}"
do
  # Calculate hash of index.html
  replaceString=($(sha1sum $1/$i/index.html))

  if [ "$DEBUG" = true ]
  then
    # If DEBUG=true in order to log the replaced files
    sed -i "s|\"\/de\/index\.html\":\s\"\(.*\)\"|\"/""$i""/index.html\": \"""$replaceString""\"|Ig;w /dev/stdout" "$1/$i/ngsw.json"
  else
    # If DEBUG=false do it without logging
    sed -i "s|""\/de\/index\.html"":\s""\(.*\)""|""/$i/index.html"": $replaceString""|Ig" "$1/$i/ngsw.json"
  fi
done

# Execute all other parameters
exec "${@:2}"
```

See it in action here: https://github.com/T-Systems-MMS/phonebook/pull/101/commits/d4e2b180d6f6674759465ec68ce7d8dbfe161228

**Comment by eimanip** [comment]
We had the same "Hash mismatch" issue in our PWA app for a different reason, which I want to share here for any other one who may encounter this issue. We have a PWA app used in production with thousands of users. It's written in Angular 9 and uses [App Shell](https://angular.io/guide/app-shell) for generating `index.html` file.

For having a better performance and load time, we were using [Angular Customer Webpack](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack) and [Compression Webpack Plugin](https://github.com/webpack-contrib/compression-webpack-plugin) plugin to make js chunks with `gzip` and `brotli` compression formats. So, they are ready in place for `nginx` to be served for any client requests for them. We intentionally wanted to generate the pre-compressed files in the build time, so they would be ready for `nginx` to be served to the clients (versus letting the `nginx` itself generate compressed files on-demand). As a result, the response time of `nginx` for such requests would be slightly better.

But here is the problem. From Angular v8.2, we weren't able to update to the latest version of Angular. I think some minor thing was changed in the way Angular builds the app (I couldn't find the root cause of it in the changelogs) which caused the generated compressed files to have a different hash after compression. So, the service worker couldn't update the app, showing this error:
`Driver state: EXISTING_CLIENTS_ONLY (Degraded due to failed initialization: Hash mismatch (cacheBustedFetchFromNetwork)`. After we disabled the "Compression Webpack Plugin" in the build pipeline and let the `nginx` itself to compress the files, the problem has been resolved. Now, we've successfully upgraded to Angular v9 without any problems with the service worker.

**Comment by mmatczak** [comment]
I have the same problem. `ngsw/state` says my `index.html` has a hash mismatch. Does anyone know how to get the file version which is being compared to the one which was taken to calculate the hash during the build?

**Comment by eimanip** [comment]
> I have the same problem. `ngsw/state` says my `index.html` has a hash mismatch. Does anyone know how to get the file version which is being compared to the one which was taken to calculate the hash during the build?

Are you changing the generated `index.html` by Angular during or after the build?

**Comment by mmatczak** [comment]
No I don't. But I let Angular CLI generate the production version on Windows while my production server runs on Linux. So thought the problem might be CR / CRLF characters.

**Comment by martinbianchi** [comment]
Uf men you save my day. I was trying to figure it out for a few days. The problem is that I'm compiling the project with the Angular CLI in a Windows machine, so the end of line is CR. My application is hosted in a Linux virtual machine so the end of line are LF. When the service worker compare the hashes it never be the same and that the reason of why doesn't work for me.

I'm pretty sure that would be a solution to still compiling and windows and make it work, but for me I'm going to left windows.

Thank you @mmatczak !!

**Comment by naveedahmed1** [comment]
Any update on this?

I am using Cloudflare which applies different optimizations on files, as a results the hash mismatch for index.html file.

**Comment by naveedahmed1** [comment]
I got the issue fixed by:

1. Run production build.
2. Minify index.html file `html-minifier -o dist/index.html dist/index.html --remove-comments --collapse-whitespace --minify-js --minify-css` 
3. Minify Service Worker `terser dist/combined-sw.js -o dist/combined-sw.js -c --comments /^##/`. .
4. Generate Service Worker config file again `ngsw-config dist ngsw-config.json`

@gkalpak @filipesilva what if Angular CLI could minify these files as well during the build process?

**Comment by gkalpak** [comment]
@naveedahmed1, maybe you could open an issue for this to be discussed on the [CLI repo](https://github.com/angular/angular-cli).

**Comment by eng-dibo** [comment]
how can I delay or rebuild 'ngsw.json' with the new hashes  after modifing the build's output files (for example minifying them with terser).

so the flow will be:
1- build
2- modify the output (minify the files, for example)
3- create ngsw with the new hashes.

**Comment by naveedahmed1** [comment]
Just run  `ngsw-config dist ngsw-config.json`

**Comment by naveedahmed1** [comment]
For convenience, I have added these to my `packages.json` file under `scripts` node:

```
    "minify-index": "html-minifier -o dist/index.html dist/index.html --remove-comments --collapse-whitespace --minify-js --minify-css",
    "concat-ngsw-fcm-worker": "concat dist/ngsw-worker.js src/messaging-sw.js -o dist/combined-sw.js",
    "minify-combined-sw": "terser dist/combined-sw.js -o dist/combined-sw.js -c --comments /^##/",
    "prepare-ngsw-config": "ngsw-config dist ngsw-config.json",
```

**Comment by memcardio** [comment]
Whilst risky, I've removed `index.html` from the `resources` array so that it would work with SSR.

**Comment by onderceylan** [comment]
Hi there, we have a use case in Angular which made us hit this issue. We build our app with string tokens in environment-settings, which are being replaced on the release pipeline based on the environments it's being deployed to. This makes one of the bundle files' hash different from what it was during the build, and therefore NGSW stops working due to hash mismatch. 

It would be great if we could manage those hashes later on or disable this check when needed.

**Comment by mich71095** [comment]
Got the same issue. Server is in windows environment(IIS) and I'm in macOS. Whenever I uploaded the build output in FileZilla, index.html, style.css, ngsw-worker.js, safety-worker.js, and worker-basic.min.js, the memory size of these files seems to increase a few bytes and other files are just the same. I'm not sure why but it could be the CR and LF as mentioned by @martinbianchi. Or it could be the sha1 algorithm on how it processes exactly on the data.

As mentioned by @naveedahmed1, I solved the hash mismatch issue by minifying the mentioned files and generate new hash table with 

./node_modules/.bin/ngsw-config ./dist/<project-name> ./ngsw-config.json

Thanks to you by the way 😀👍. 

Also bypassing the hash checker could result to unwanted behavior or broken app.

**Comment by AaronDovTurkel** [comment]
Hi, this is a great thread read!
I have a tokenizer in the deploy stage of my pipeline that replaces and env variable.
It seems that this should cause an issue for the SW. Is that correct? Is there a workaround?
(the project is already built at this point)

**Comment by gkalpak** [comment]
@AaronDovTurkel, yes, it most probably will be a problem (i.e. it depends on whether the modified files are included in the `ngsw.json` manifest).
A possible way to work around that would be to re-generate `ngsw.json` after modifying the files with something like: `node_modules/.bin/ngsw-config dist/your-app ngsw-config.json`

**Comment by AaronDovTurkel** [comment]
@gkalpak right the only thing is we created a release time var with the tokenizer in our release pipeline. The app is already built and at this point we only have access to the artifact (aka the dist).

**Comment by hiepxanh** [comment]
hello, my service worker have hash missmatch problem, I guess there is facebook causing this error. I have question:
1) how can I skip facebook url cache on SW?
2) how to force reload? I mean force update, clear service worker itself

**Comment by Dador** [comment]
I have the same issue, in my case it caused by MITM proxy that client using to block ads ([AdGuard for Android](https://adguard.com/en/adguard-android/overview.html)). That MITM proxy adds `<script>...</script>` to `index.html` and of course hash doesn't matches after such manipulations. Because it's happens on client side there is no way to predict new hash.

Maybe it's good idea to add the hash inside file contents? That way we can confirm that we dowloaded latest version of the file and still allow some modifications (especially the ones we can't control).

**Comment by antal-horvath** [comment]
@mich71095 Thanks for your comment. I had the same issue and changing the transfer type to binary solved the problem.

Whenever I copied over e.g. index.html and main.js to the web server, the ServiceWorkers did not register updates, and when I copied a new ngsw.json, then I received the VersionInstallationFailedEvent due to hash mismatches.

**Comment by webdeveloperninja** [comment]
Our release pipeline does a token replace on a configuration file in the assets folder to handle build once deploy to multiple environments. Updating the file changes the sha hash in the `ngsw.json`  `hashTables` object. This node script ended up working after running this after the token replace task. This issue helped me solve my problem. Was stuck for a good 3 days.

```
const crypto = require('crypto');
const fs = require('fs');

var getSHA1ofJSON = function(input){
  const hash = crypto.createHash('sha1');
  hash.update(input);
  
  return hash.digest('hex');
}

const args = getArgs();

const pathToSettings = args.pathToSettings;
const pathToNgswJson = args.pathToNgswJson;

let ngsw = JSON.parse(fs.readFileSync(pathToNgswJson));

const fileBuffer = fs.readFileSync(pathToSettings);
const sha1 = getSHA1ofJSON(fileBuffer);

ngsw.hashTable['/assets/settings.tokenize.json'] = sha1;

fs.writeFileSync(pathToNgswJson, JSON.stringify(ngsw));
```

**Comment by yoazerty** [comment]
Had the same issue as @hqrd with Dynatrace adding script tag in my index.html file, causing the Service Worker hash mismatch problem... searched for hours...

**Comment by bjornharvold** [comment]
In my case, when I turned off Cloudflare content optimizations, Service Worker started to work again like magic; just like @manfredsteyer stated in the ticket mentioned above.

<img width="1046" alt="Screenshot 2023-07-10 at 18 45 15" src="https://github.com/angular/angular/assets/641791/af16a295-fe69-4864-a0fa-7da30a2c43af">

**Comment by hqrd** [comment]
> > > @hqrd if you can get the script tag in the build then,as @gkalpak said, you can regenerate the hash table after the script is injected.
> > 
> > 
> > I'm doing that already for other files with environment variables, but it doesn't work in this case because the script tag is added after my app is started, and has some changing variables.
> 
> @hqrd Did you ever solve your issue?

Yes, by playing with the `/node_modules/.bin/ngsw-config` in the Dockerfile, to track updated env variables, instead of the index.html, and update the ngsw-config accordingly yes. Basically what @webdeveloperninja commented.

**Comment by JeanMeche** [comment]
With https://github.com/angular/angular/pull/63364 we've announced that the service-worker package in now in maintenance mode and will not receive feature PRs and feature requests.

The message is also visible here: https://angular.dev/ecosystem/service-workers
